### DISCUSSION TITLE
Why Halo2?

### DISCUSSION THREAD

**Discussion by hanbu97** [z2o-k7e/halo2-step-by-step#6] [Q&A]
背景：之前只用circom+snarkjs写过零知识证明应用。

今天看到了一篇文章，对比了几种现有的zk proving system

https://blog.celer.network/2023/08/04/the-pantheon-of-zero-knowledge-proof-development-frameworks/

产生了一个疑问：
从生态来讲，从生态，证明生成速度，占用内存大小，证明大小，验证消耗gas来讲，几种方案各有优劣，似乎Halo2不是最优秀的那一个。

所以为什么大家推崇Halo2呢？它的优势在哪里。

**Comment by luckyyang** [comment]
Halo2 生态成熟，被验证安全可用

**Comment by Demian101** [comment]
A Q&A from halo2 discord for you :

Q
**according to you is halo2 better than **circom** and why?**
A
I think there's no better or worse. They serve the same purpose but apply different techniques.

While in halo2 you have the power and expressiveness of Rust to help you, you also pay the prize with more verbosity.

Instead with Circom, you don't have that expressiveness and instead you have concise instructions which are purely the circuit and not rust-only stuff.

Depends on the taste and person. I can't say that something is categorically better or worse. 


------ 

another version :

### Comparative Study of Circom, Arkworks, and ZoKrates

![](http://imagesoda.oss-cn-beijing.aliyuncs.com/Sodaoo/2023-07-30-013847.png)

Finally, note that all of these tools **essentially output an R1CS**, or more specific types of it like Plonk or AIR. So, within that process, all these tools share quite a bit of commontechniques. With that in mind, a library to create ZKP languages actually exists: [Circ](https://github.com/circify/circ).

You can also check out [https://zkp.science/](https://zkp.science/) which has a great coverage of tools, as well as the development of ZK theory.

**Reply by Demian101** [reply]
another Q&A from @dajuguan  for you:

为什么 circom 中的 plonk 要求使用 r1cs 约束, 而 halo2 的约束可以任意次的多项式?

@Po 首先plonk是有不同版本的，最初的 plonk 论文是vanilla plonk(只包含加法门和乘法门)，再到后来的turboplonk等(支持自定义门)发展的plonkish。我的理解是circom本来设计之初就是为了支持R1CS，只不过 r1cs 可以很容易的转换为vanilla plonk; halo2一开始设计就是针对plonkish + lookup的。所以circom要支持plonkish整个代码设计就要重新调整

**Reply by hanbu97** [reply]
主要是Halo2和

Plonky2, Boojum, Starky

这三个相比

**Comment by Ah-lam** [comment]


**Reply by pochenai** [reply]
整体写的很好，但是感觉不太准确，halo2其实不是个算法，他是一个模块化的框架。Halo2的证明系统可以是IPA, Plonk, Stark，所以上述表格对比证明时间应该加上Halo2/IPA的限定。我觉得Halo2有个很强大的地方就是模块化做的特别好，将前端算数化、后端证明系统进行了明确的分层，使得开发者很容易进行定制，所以出现了Halo2 Wrong/PSE/Scroll/Axiom等各种Halo2的版本；并且Halo2在早期plonky等框架出来之前是Zcash在生产级别使用过的框架，具有工业级别的稳定性，当时只有Groth16在Filecoin这种大型项目上使用了，但是G16又需要trusted per circuit。

**Reply by kurtpan666** [reply]
不想打击积极性，但如果这都是你自己理解的，那该回答问题其实很严重。似是而非的理解往往还不如不理解。

**Reply by amimibear** [reply]
Spartan底层所使用的polynomial commitment scheme是Hyrax-commit，sum-check protocol是Spartan在pIOP(MIP-based)部分用到的，和pcs无关
